worker/src/artifact.rs

```diff
@@ -259,905 +273,455 @@ impl ArtifactService for ArtifactServer {
-                        "failed to remove workspace: {:?}",
-                        err
-                    ))))
-                    .await
-                {
-                    error!("failed to send error: {:?}", err);
-                }
+    if source_archive_path.exists() {
+        send_message(
+            tx,
+            format!("caching source: {}-{}", source.name, source.hash),
+        )
+        .await?;
+
+        if let Err(err) = create_dir_all(&source_cache_path).await {
+            return Err(Status::internal(format!(
+                "failed to create source path: {:?}",
+                err
+            )));
+        }
+
+        if let Err(err) = unpack_zstd(&source_cache_path, &source_archive_path).await {
+            return Err(Status::internal(format!(
+                "failed to unpack source archive: {:?}",
+                err
+            )));
+        }
+
+        let source_cache_files = get_file_paths(&source_cache_path, vec![], vec![])
+            .map_err(|err| Status::internal(format!("failed to get source files: {:?}", err)))?;
+
+        send_message(
+            tx,
+            format!("copying source: {}-{}", source.name, source.hash),
+        )
+        .await?;
 
-                return;
+        let workspace_source_files = copy_files(
+            &source_cache_path,
+            source_cache_files,
+            &workspace_source_path,
+        )
+        .await
+        .map_err(|err| Status::internal(format!("failed to copy source files: {:?}", err)))?;
+
+        for path in workspace_source_files.iter() {
+            if let Err(err) = set_timestamps(path).await {
+                return Err(Status::internal(format!(
+                    "failed to sanitize output files: {:?}",
+                    err
+                )));
             }
+        }
 
-            // Remove lock file
-
-            if let Err(err) = remove_file(&lock_path).await {
-                if let Err(err) = tx
-                    .send(Err(Status::internal(format!(
-                        "failed to remove lock file: {:?}",
-                        err
-                    ))))
-                    .await
-                {
-                    error!("failed to send error: {:?}", err);
-                }
+        return Ok(());
+    }
+
+    send_message(
+        tx,
+        format!("pulling source: {}-{}", source.name, source.hash),
+    )
+    .await?;
+
+    let pull_request = RegistryRequest {
+        hash: source.hash.clone(),
+        name: source.name.clone(),
+        kind: RegistryKind::ArtifactSource as i32,
+    };
+
+    let response = registry_client.pull(pull_request).await.map_err(|status| {
+        if status.code() != NotFound {
+            Status::internal(format!("failed to pull source archive: {:?}", status))
+        } else {
+            // FIXME: What error should be returned in this case?
+            Status::internal(format!("failed to pull source archive: {:?}", status))
+        }
```

Comment:
There is no need for this to handle as an if statement anymore. If the artifact service finds a source in the configuration but can't pull it there was a problem with the SDK or CLI pushing it.